# pandas

Now that we know how to get our data in, let's look a little more at `pandas` and what it can do. To do this, let's bring in some data from Zillow.

The data frame we will be working with contains data from Zillow.com, the housing web site, on the number of homes, the median value of homes, etc. at-risk for flooding due to rising sea levels.

You can find many different Zillow data sets [here](https://www.zillow.com/research/data/). The “uw” here stands for “under water” in a very literal sense. I first thought they were talking about mortgages!

As mentioned, take a look at [Chapter 5](https://wesmckinney.com/book/pandas-basics.html) of *Python for Data Analysis* for an introduction to `pandas`. 

In [1]:
import numpy as np
import pandas as pd

uw = pd.read_csv('https://github.com/aaiken1/fin-data-analysis-python/raw/main/data/zestimatesAndCutoffs_byGeo_uw_2017-10-10_forDataPage.csv')

We are going to use `read_csv` to get this data off of my Github page. `pandas` also comes with `read_excel`, `read_json`, `read_html`, and other ways to get formatted data into Python.

## Looking at our data

Let's peak at this data. You can do that below, or using the `.info` method. We'll get variable names, the number of non-missing (null) observations, and the type of data. One number variable, `UWHomes_Tier2`, got brought in as an integer, rather than a float. When you look at the data below, notice how that one column doesn't have a `.0` on any of the numbers. Integers can't have decimal places.

In [2]:
uw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2610 entries, 0 to 2609
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   RegionType                    2610 non-null   object 
 1   RegionName                    2610 non-null   object 
 2   StateName                     2609 non-null   object 
 3   MSAName                       1071 non-null   object 
 4   AllHomes_Tier1                2610 non-null   float64
 5   AllHomes_Tier2                2610 non-null   float64
 6   AllHomes_Tier3                2610 non-null   float64
 7   AllHomes_AllTiers             2610 non-null   float64
 8   UWHomes_Tier1                 2610 non-null   float64
 9   UWHomes_Tier2                 2610 non-null   int64  
 10  UWHomes_Tier3                 2610 non-null   float64
 11  UWHomes_AllTiers              2610 non-null   float64
 12  UWHomes_TotalValue_Tier1      2610 non-null   float64
 13  UWH

Just to show you, I'm going to reimport that data, but force that one variable to also come in as the same variable type. I don't want anything unexpected to happen, so let's get all numerics as floats.

In [3]:
uw = pd.read_csv('https://github.com/aaiken1/fin-data-analysis-python/raw/main/data/zestimatesAndCutoffs_byGeo_uw_2017-10-10_forDataPage.csv', dtype={'UWHomes_Tier2':np.float64})
uw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2610 entries, 0 to 2609
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   RegionType                    2610 non-null   object 
 1   RegionName                    2610 non-null   object 
 2   StateName                     2609 non-null   object 
 3   MSAName                       1071 non-null   object 
 4   AllHomes_Tier1                2610 non-null   float64
 5   AllHomes_Tier2                2610 non-null   float64
 6   AllHomes_Tier3                2610 non-null   float64
 7   AllHomes_AllTiers             2610 non-null   float64
 8   UWHomes_Tier1                 2610 non-null   float64
 9   UWHomes_Tier2                 2610 non-null   float64
 10  UWHomes_Tier3                 2610 non-null   float64
 11  UWHomes_AllTiers              2610 non-null   float64
 12  UWHomes_TotalValue_Tier1      2610 non-null   float64
 13  UWH

OK, that seems better. All of these functions that we're using have many different options. The only way to learn about them is to play around, read help documents, and search online for solutions to our problems.

There's a concept in data analysis called **tidy data**. The characteristics of tidy data are:

- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.

```{figure} ../images/04-tidy.png
---
name: 04-tidy.png
align: center
---
Tidy data.
```

The first thing to note about this data is that there are several tiers. Some numbers are at the national level, some at the region, some at the state, and some at the MSA. **Tidy data** refers to one row, one observation, across different variables (columns) and, possibly, across time. When you have multiple observations of the same firm, state, person, etc. over time, we call that panel data. Panel data is tidy too.

Note that some RegionTypes (e.g. Nation, State) are actually summaries of other rows. This is bit confusing and we need to be mindful of this different levels when using the data.

Before we get to that, just a few more methods to look at the data.

First, let's look at the first five rows.

In [4]:
uw.head()

RegionType     RegionName    StateName MSAName  AllHomes_Tier1  \
0     Nation  United States          NaN     NaN      35461549.0   
1      State        Alabama      Alabama     NaN        546670.0   
2      State     California   California     NaN       3060171.0   
3      State    Connecticut  Connecticut     NaN        333904.0   
4      State       Delaware     Delaware     NaN        102983.0   

   AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
0      35452941.0      35484532.0        106399022.0       594672.0   
1        520247.0        491300.0          1558217.0         2890.0   
2       3076238.0       3169584.0          9305993.0         8090.0   
3        334688.0        336254.0          1004846.0         4434.0   
4        127573.0        127983.0           358539.0         4105.0   

   UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  UWHomes_TotalValue_AllTiers  \
0       542681.0  ...              5.974490e+11                 9.159540e+11   
1         2766.0  ...              2.583039e+09                 3.578203e+09   
2        14266.0  ...              2.666638e+10                 4.449536e+10   
3         3807.0  ...              8.689480e+09                 1.152170e+10   
4         2574.0  ...              1.498013e+09                 2.801847e+09   

   UWHomes_MedianValue_AllTiers  AllHomes_Tier1_ShareUW  \
0                      310306.0                0.016769   
1                      270254.0                0.005287   
2                      860373.0                0.002644   
3                      442036.0                0.013279   
4                      241344.5                0.039861   

   AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  AllHomes_AllTiers_ShareUW  \
0                0.015307                0.020458                   0.017512   
1                0.005317                0.010271                   0.006868   
2                0.004637                0.005611                   0.004313   
3                0.011375                0.021677                   0.015455   
4                0.020177                0.025675                   0.027793   

   UWHomes_ShareInTier1  UWHomes_ShareInTier2  UWHomes_ShareInTier3  
0              0.319149              0.291246              0.389605  
1              0.270043              0.258456              0.471501  
2              0.201550              0.355415              0.443035  
3              0.285512              0.245138              0.469350  
4              0.411942              0.258304              0.329754  

[5 rows x 24 columns]

`NaN` is how `pandas` shows missing values. So, that first row is `RegionType == 'Nation'` and `RegionName == 'UnitedStates'`. `StateName` and `MSAName` are undefined, since this row is **all of the data added up**. We usually don't want to have summary statistics in the same DataFrame as underlying data! What a mess!

We can also look at the last five rows.

In [5]:
uw.tail()

RegionType RegionName   StateName MSAName  AllHomes_Tier1  \
2605        Zip      98592  Washington     NaN           464.0   
2606        Zip      98595  Washington     NaN           558.0   
2607        Zip      98612  Washington     NaN           365.0   
2608        Zip      32081     Florida     NaN          1318.0   
2609        Zip      33578     Florida     NaN          5975.0   

      AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
2605           470.0           493.0             1427.0            8.0   
2606           571.0           598.0             1727.0          119.0   
2607           376.0           409.0             1150.0           91.0   
2608          1328.0          1350.0             3996.0           91.0   
2609          4537.0          3232.0            13744.0            8.0   

      UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  \
2605           29.0  ...                73407109.0   
2606          104.0  ...                13761194.0   
2607           79.0  ...                14406737.0   
2608           42.0  ...                91112117.0   
2609            6.0  ...                56852018.0   

      UWHomes_TotalValue_AllTiers  UWHomes_MedianValue_AllTiers  \
2605                   82715412.0                      436490.5   
2606                   40365702.0                      131441.0   
2607                   50284001.0                      230495.0   
2608                  132505542.0                      385196.0   
2609                   59115506.0                      363355.0   

      AllHomes_Tier1_ShareUW  AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  \
2605                0.017241                0.061702                0.273834   
2606                0.213262                0.182137                0.095318   
2607                0.249315                0.210106                0.092910   
2608                0.069044                0.031627                0.097778   
2609                0.001339                0.001322                0.038366   

      AllHomes_AllTiers_ShareUW  UWHomes_ShareInTier1  UWHomes_ShareInTier2  \
2605                   0.120533              0.046512              0.168605   
2606                   0.162131              0.425000              0.371429   
2607                   0.180870              0.437500              0.379808   
2608                   0.066316              0.343396              0.158491   
2609                   0.010041              0.057971              0.043478   

      UWHomes_ShareInTier3  
2605              0.784884  
2606              0.203571  
2607              0.182692  
2608              0.498113  
2609              0.898551  

[5 rows x 24 columns]

Looking at this data, I think it really has four levels that describe the numbers. The columns `RegionType`, `RegionName`, `StateName`, and `MSAName` define unique rows of data. Again, note that these values can be missing. For example, row 2606 is Zip, 98595, Washington, but no `MSAName`. Why? MSA stands for Metropolitan Statistical Area and covers multiple zip codes. This data is **not tidy**.

Let’s think about the other variable definitions a bit. Zillow places homes into three tiers based on the home’s estimated market value, with Tier3 being the highest value. `AllHomes` refers to all homes in that tier or across all tiers. `UWHomes` refers to homes that are at risk for flooding. Note that there are some variables that are counts, some that are dollar values, and others that are percentages.



## Selecting our data

You access columns in `pandas` like this, using `[]` and `''`. A column is a variable in our data set.

In [6]:
uw['RegionType']

0       Nation
1        State
2        State
3        State
4        State
         ...  
2605       Zip
2606       Zip
2607       Zip
2608       Zip
2609       Zip
Name: RegionType, Length: 2610, dtype: object

But, wait, you can also do this!

In [7]:
uw.RegionType

0       Nation
1        State
2        State
3        State
4        State
         ...  
2605       Zip
2606       Zip
2607       Zip
2608       Zip
2609       Zip
Name: RegionType, Length: 2610, dtype: object

Notice how as you're typing, VS Code is trying to autocomplete for you? Helpful!

You can also do multiple columns. Use a `,` to separate the column names in `[]`.


In [8]:
uw[['RegionType', 'RegionName', 'UWHomes_Tier2']]

RegionType     RegionName  UWHomes_Tier2
0        Nation  United States       542681.0
1         State        Alabama         2766.0
2         State     California        14266.0
3         State    Connecticut         3807.0
4         State       Delaware         2574.0
...         ...            ...            ...
2605        Zip          98592           29.0
2606        Zip          98595          104.0
2607        Zip          98612           79.0
2608        Zip          32081           42.0
2609        Zip          33578            6.0

[2610 rows x 3 columns]

This is looking a lot like dealing with `lists`. You can also pull out just certain rows. We used `head` and `tail` above to do this in an "automated" way.

In [9]:
uw[0:10]

RegionType     RegionName      StateName MSAName  AllHomes_Tier1  \
0     Nation  United States            NaN     NaN      35461549.0   
1      State        Alabama        Alabama     NaN        546670.0   
2      State     California     California     NaN       3060171.0   
3      State    Connecticut    Connecticut     NaN        333904.0   
4      State       Delaware       Delaware     NaN        102983.0   
5      State        Florida        Florida     NaN       2391491.0   
6      State        Georgia        Georgia     NaN       1003890.0   
7      State         Hawaii         Hawaii     NaN        132718.0   
8      State      Louisiana      Louisiana     NaN        439606.0   
9      State  Massachusetts  Massachusetts     NaN        610378.0   

   AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
0      35452941.0      35484532.0        106399022.0       594672.0   
1        520247.0        491300.0          1558217.0         2890.0   
2       3076238.0       3169584.0          9305993.0         8090.0   
3        334688.0        336254.0          1004846.0         4434.0   
4        127573.0        127983.0           358539.0         4105.0   
5       2435575.0       2456637.0          7283703.0       313090.0   
6        986416.0        920843.0          2911149.0         5433.0   
7        132760.0        139397.0           404875.0        16716.0   
8        424665.0        417773.0          1282044.0        25678.0   
9        611776.0        611920.0          1834074.0        16830.0   

   UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  UWHomes_TotalValue_AllTiers  \
0       542681.0  ...              5.974490e+11                 9.159540e+11   
1         2766.0  ...              2.583039e+09                 3.578203e+09   
2        14266.0  ...              2.666638e+10                 4.449536e+10   
3         3807.0  ...              8.689480e+09                 1.152170e+10   
4         2574.0  ...              1.498013e+09                 2.801847e+09   
5       263276.0  ...              2.782440e+11                 4.163480e+11   
6         6904.0  ...              6.102691e+09                 9.565510e+09   
7        12536.0  ...              1.134778e+10                 2.516682e+10   
8        23280.0  ...              6.755054e+09                 1.309180e+10   
9        16491.0  ...              3.436795e+10                 4.988792e+10   

   UWHomes_MedianValue_AllTiers  AllHomes_Tier1_ShareUW  \
0                      310306.0                0.016769   
1                      270254.0                0.005287   
2                      860373.0                0.002644   
3                      442036.0                0.013279   
4                      241344.5                0.039861   
5                      276690.0                0.130918   
6                      295405.0                0.005412   
7                      500666.0                0.125951   
8                      149888.0                0.058411   
9                      582126.0                0.027573   

   AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  AllHomes_AllTiers_ShareUW  \
0                0.015307                0.020458                   0.017512   
1                0.005317                0.010271                   0.006868   
2                0.004637                0.005611                   0.004313   
3                0.011375                0.021677                   0.015455   
4                0.020177                0.025675                   0.027793   
5                0.108096                0.135070                   0.124687   
6                0.006999                0.010676                   0.007615   
7                0.094426                0.051658                   0.090035   
8                0.054820                0.060408                   0.057872   
9                0.026956                0.040749                   0.031763   

   UWHomes_ShareInTier1  UWHomes_Shar

Again, notice how Python starts counting from 0. That 0, 1, 2... shows the current `index` for this `DataFrame`. We could change this and will later.

`pandas` has two **location** methods: `loc` and `iloc`. They are going to let us select parts of our data. The main difference: `loc` is labelled based and `iloc` is location based. Since our columns have labels (our rows do not), we can use `loc` to specify columns by name. If we created an **index** for this data, then our rows would have labels. We'll do that later. Any index for this data will be complicated, since no one column uniquely identifies anything!

You can read more about the two methods (here)[https://towardsdatascience.com/how-to-use-loc-and-iloc-for-selecting-data-in-pandas-bd09cb4c3d79]. 

We'll start with `loc`. Again, our rows don't have labels or an index, so we can use the number of the row, even with `loc`. We can pull just the first row of data.

In [10]:
uw.loc[0]

RegionType                              Nation
RegionName                       United States
StateName                                  NaN
MSAName                                    NaN
AllHomes_Tier1                      35461549.0
AllHomes_Tier2                      35452941.0
AllHomes_Tier3                      35484532.0
AllHomes_AllTiers                  106399022.0
UWHomes_Tier1                         594672.0
UWHomes_Tier2                         542681.0
UWHomes_Tier3                         725955.0
UWHomes_AllTiers                     1863308.0
UWHomes_TotalValue_Tier1        122992000000.0
UWHomes_TotalValue_Tier2        195513000000.0
UWHomes_TotalValue_Tier3        597449000000.0
UWHomes_TotalValue_AllTiers     915954000000.0
UWHomes_MedianValue_AllTiers          310306.0
AllHomes_Tier1_ShareUW                0.016769
AllHomes_Tier2_ShareUW                0.015307
AllHomes_Tier3_ShareUW                0.020458
AllHomes_AllTiers_ShareUW             0.017512
UWHomes_Share

We can pull the first ten rows and just three columns. But, now we are going to use the column names. And - notice how `0:9` actually gets us `0:9`! Because we are using `loc` we include the last item in the range.

In [11]:
uw.loc[0:9,['RegionType', 'RegionName', 'UWHomes_Tier2']]

RegionType     RegionName  UWHomes_Tier2
0     Nation  United States       542681.0
1      State        Alabama         2766.0
2      State     California        14266.0
3      State    Connecticut         3807.0
4      State       Delaware         2574.0
5      State        Florida       263276.0
6      State        Georgia         6904.0
7      State         Hawaii        12536.0
8      State      Louisiana        23280.0
9      State  Massachusetts        16491.0

We can pull a range for columns too. Again, notice that the column name range is inclusive of the last column, unlike with slicing by element number.

In [12]:
uw.loc[0:9,'RegionType':'UWHomes_Tier2']

RegionType     RegionName      StateName MSAName  AllHomes_Tier1  \
0     Nation  United States            NaN     NaN      35461549.0   
1      State        Alabama        Alabama     NaN        546670.0   
2      State     California     California     NaN       3060171.0   
3      State    Connecticut    Connecticut     NaN        333904.0   
4      State       Delaware       Delaware     NaN        102983.0   
5      State        Florida        Florida     NaN       2391491.0   
6      State        Georgia        Georgia     NaN       1003890.0   
7      State         Hawaii         Hawaii     NaN        132718.0   
8      State      Louisiana      Louisiana     NaN        439606.0   
9      State  Massachusetts  Massachusetts     NaN        610378.0   

   AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
0      35452941.0      35484532.0        106399022.0       594672.0   
1        520247.0        491300.0          1558217.0         2890.0   
2       3076238.0       3169584.0          9305993.0         8090.0   
3        334688.0        336254.0          1004846.0         4434.0   
4        127573.0        127983.0           358539.0         4105.0   
5       2435575.0       2456637.0          7283703.0       313090.0   
6        986416.0        920843.0          2911149.0         5433.0   
7        132760.0        139397.0           404875.0        16716.0   
8        424665.0        417773.0          1282044.0        25678.0   
9        611776.0        611920.0          1834074.0        16830.0   

   UWHomes_Tier2  
0       542681.0  
1         2766.0  
2        14266.0  
3         3807.0  
4         2574.0  
5       263276.0  
6         6904.0  
7        12536.0  
8        23280.0  
9        16491.0

Now, let's try `iloc`. I'll pull the exact same data as above. The numbers used are going to work like slicing. We want to use `0:10` for the rows, because we want the first 10 rows. Same for the columns. Inclusive of the first number, but up to and excluding the second.

In [13]:
uw.iloc[0:10,0:10]

RegionType     RegionName      StateName MSAName  AllHomes_Tier1  \
0     Nation  United States            NaN     NaN      35461549.0   
1      State        Alabama        Alabama     NaN        546670.0   
2      State     California     California     NaN       3060171.0   
3      State    Connecticut    Connecticut     NaN        333904.0   
4      State       Delaware       Delaware     NaN        102983.0   
5      State        Florida        Florida     NaN       2391491.0   
6      State        Georgia        Georgia     NaN       1003890.0   
7      State         Hawaii         Hawaii     NaN        132718.0   
8      State      Louisiana      Louisiana     NaN        439606.0   
9      State  Massachusetts  Massachusetts     NaN        610378.0   

   AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
0      35452941.0      35484532.0        106399022.0       594672.0   
1        520247.0        491300.0          1558217.0         2890.0   
2       3076238.0       3169584.0          9305993.0         8090.0   
3        334688.0        336254.0          1004846.0         4434.0   
4        127573.0        127983.0           358539.0         4105.0   
5       2435575.0       2456637.0          7283703.0       313090.0   
6        986416.0        920843.0          2911149.0         5433.0   
7        132760.0        139397.0           404875.0        16716.0   
8        424665.0        417773.0          1282044.0        25678.0   
9        611776.0        611920.0          1834074.0        16830.0   

   UWHomes_Tier2  
0       542681.0  
1         2766.0  
2        14266.0  
3         3807.0  
4         2574.0  
5       263276.0  
6         6904.0  
7        12536.0  
8        23280.0  
9        16491.0

We can also select columns in various locations, like above.

In [14]:
uw.iloc[0:10,[0, 1, 9]]

RegionType     RegionName  UWHomes_Tier2
0     Nation  United States       542681.0
1      State        Alabama         2766.0
2      State     California        14266.0
3      State    Connecticut         3807.0
4      State       Delaware         2574.0
5      State        Florida       263276.0
6      State        Georgia         6904.0
7      State         Hawaii        12536.0
8      State      Louisiana        23280.0
9      State  Massachusetts        16491.0

Let's try **filtering** our data and keeping just the MSA observations. The text calls this **complex selection**. See pg. 132.

Here's some [help on filtering](https://datagy.io/filter-pandas/) data with `pandas`.

In [15]:
uw[uw['RegionType'] == 'MSA'].head(15)

RegionType               RegionName      StateName  \
251        MSA             Aberdeen, WA     Washington   
252        MSA              Astoria, OR         Oregon   
253        MSA  Egg Harbor Township, NJ     New Jersey   
254        MSA            Baltimore, MD       Maryland   
255        MSA        East Falmouth, MA  Massachusetts   
256        MSA          Baton Rouge, LA      Louisiana   
257        MSA             Bay City, TX          Texas   
258        MSA             Beaumont, TX          Texas   
259        MSA           Bellingham, WA     Washington   
260        MSA               Boston, MA  Massachusetts   
261        MSA            Bremerton, WA     Washington   
262        MSA           Bridgeport, CT    Connecticut   
263        MSA          Brownsville, TX          Texas   
264        MSA            Brunswick, GA        Georgia   
265        MSA            Cambridge, MD       Maryland   

                     MSAName  AllHomes_Tier1  AllHomes_Tier2  AllHomes_Tier3  \
251             Aberdeen, WA         11606.0         11751.0         11749.0   
252              Astoria, OR          4745.0          5228.0          5568.0   
253  Egg Harbor Township, NJ         34185.0         34117.0         34088.0   
254            Baltimore, MD        298318.0        312603.0        316502.0   
255        East Falmouth, MA         45345.0         45358.0         45258.0   
256          Baton Rouge, LA         78815.0         79581.0         78731.0   
257             Bay City, TX          4161.0          4252.0          4030.0   
258             Beaumont, TX         49842.0         47829.0         44616.0   
259           Bellingham, WA         29239.0         30413.0         30956.0   
260               Boston, MA        409592.0        409966.0        410140.0   
261            Bremerton, WA         27865.0         27856.0         28119.0   
262           Bridgeport, CT         84535.0         84952.0         85028.0   
263          Brownsville, TX         47084.0         37642.0         35503.0   
264            Brunswick, GA         14827.0         14885.0         15532.0   
265            Cambridge, MD          6193.0          6251.0          6118.0   

     AllHomes_AllTiers  UWHomes_Tier1  UWHomes_Tier2  ...  \
251            35106.0         4119.0         2450.0  ...   
252            15541.0          811.0          376.0  ...   
253           102390.0         9116.0         9575.0  ...   
254           927423.0         2293.0         2910.0  ...   
255           135961.0         1651.0         1395.0  ...   
256           237127.0         2214.0         1904.0  ...   
257            12443.0          195.0          351.0  ...   
258           142287.0         1182.0         1043.0  ...   
259            90608.0         2033.0          812.0  ...   
260          1229698.0        15948.0        15210.0  ...   
261            83840.0          206.0          245.0  ...   
262           254515.0         2326.0         1895.0  ...   
263           120229.0         1621.0         1430.0  ...   
264            45244.0         2532.0         2924.0  ...   
265            18562.0          757.0         1119.0  ...   

     UWHomes_TotalValue_Tier3  UWHomes_TotalValue_AllTiers  \
251              1.734113e+08                 7.791588e+08   
252              9.138231e+07                 3.620604e+08   
253              5.576646e+09                 1.030631e+10   
254              3.698818e+09                 5.153664e+09   
255              4.205784e+09                 5.349099e+09   
256              5.392566e+08                 1.219648e+09   
257              1.175738e+08                 1.964820e+08   
258              1.466850e+08                 3.249066e+08   
259              7.000724e+08                 1.184436e+09   
260              2.851453e+10                 4.266696e+10   
261              2.076948e+09                 2.275970e+09   
262              6.041231e+09                 7.812512e+09   
263        

How about just North Carolina MSAs? I'll use the bigger DataFrame and just show the observations, rather than create a new DataFrame.

In [16]:
uw[(uw['RegionType'] == 'MSA') & (uw['StateName'] == 'North Carolina')].head(15)

RegionType            RegionName       StateName               MSAName  \
274        MSA    Elizabeth City, NC  North Carolina    Elizabeth City, NC   
277        MSA        Greenville, NC  North Carolina        Greenville, NC   
288        MSA      Jacksonville, NC  North Carolina      Jacksonville, NC   
292        MSA  Kill Devil Hills, NC  North Carolina  Kill Devil Hills, NC   
297        MSA     Morehead City, NC  North Carolina     Morehead City, NC   
302        MSA          New Bern, NC  North Carolina          New Bern, NC   
334        MSA        Washington, NC  North Carolina        Washington, NC   
336        MSA        Wilmington, NC  North Carolina        Wilmington, NC   

     AllHomes_Tier1  AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  \
274          5887.0          5882.0          5868.0            17637.0   
277         22042.0         19844.0         18836.0            60722.0   
288         17623.0         19413.0         17989.0            55025.0   
292          3194.0          3213.0          3486.0             9893.0   
297         13555.0         13843.0         12927.0            40325.0   
302         13637.0         13409.0         12440.0            39486.0   
334          1289.0          1069.0          1043.0             3401.0   
336         37100.0         40034.0         37475.0           114609.0   

     UWHomes_Tier1  UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  \
274         2198.0         1364.0  ...              4.784827e+08   
277           71.0           34.0  ...              6.786239e+06   
288          792.0         1123.0  ...              6.370269e+08   
292         1461.0         1584.0  ...              9.128899e+08   
297         3195.0         3308.0  ...              2.316937e+09   
302         1364.0         1138.0  ...              5.442292e+08   
334          255.0          201.0  ...              5.513269e+07   
336         3367.0         3339.0  ...              4.265842e+09   

     UWHomes_TotalValue_AllTiers  UWHomes_MedianValue_AllTiers  \
274                 8.833370e+08                      138246.0   
277                 1.776378e+07                       99179.5   
288                 1.029168e+09                      264070.5   
292                 1.650730e+09                      280409.0   
297                 3.449861e+09                      244691.0   
302                 8.553945e+08                      163643.0   
334                 9.990351e+07                      116902.5   
336                 6.761845e+09                      383067.0   

     AllHomes_Tier1_ShareUW  AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  \
274                0.373365                0.231894                0.283401   
277                0.003221                0.001713                0.001221   
288                0.044941                0.057848                0.083329   
292                0.457420                0.492997                0.542456   
297                0.235706                0.238966                0.359712   
302                0.100022                0.084868                0.132235   
334                0.197828                0.188026                0.182167   
336                0.090755                0.083404                0.127578   

     AllHomes_AllTiers_ShareUW  UWHomes_ShareInTier1  UWHomes_ShareInTier2  \
274                   0.296252              0.420670              0.261053   
277                   0.002108              0.554688              0.265625   
288                   0.062045              0.231986              0.328940   
292                   0.498939              0.295989              0.320908   
297                   0.276578              0.286470              0.296602   
302                   0.105025              0.328912              0.274415   
334                   0.189944              0.394737              0.311146   
336                   0.100228              0.293114              0.290676   

     UWHome

Notice how I needed to put `()` around each condition. Python is testing to see if each of these are `True` or `False` and then filtering our data accordingly. Pgs. 132 - 135 of our text show you more about this type of **boolean** filtering.

## Dealing with indices

Let's go back to the stock data and work with indices a bit more. Remember how that one had the `Date` as the index?

In [17]:
prices = pd.read_csv('https://github.com/aaiken1/fin-data-analysis-python/raw/main/data/tr_eikon_eod_data.csv',
                      index_col=0, parse_dates=True)

prices.index

DatetimeIndex(['2010-01-01', '2010-01-04', '2010-01-05', '2010-01-06',
               '2010-01-07', '2010-01-08', '2010-01-11', '2010-01-12',
               '2010-01-13', '2010-01-14',
               ...
               '2018-06-18', '2018-06-19', '2018-06-20', '2018-06-21',
               '2018-06-22', '2018-06-25', '2018-06-26', '2018-06-27',
               '2018-06-28', '2018-06-29'],
              dtype='datetime64[ns]', name='Date', length=2216, freq=None)

Now, let's try `loc`, but using our `dtype=datetime` index. We'll pull in the SPX and VIX for just January 2010. 

In [18]:
prices.loc['2010-01-01':'2010-01-31',['.SPX', '.VIX']]

.SPX   .VIX
Date                      
2010-01-01      NaN    NaN
2010-01-04  1132.99  20.04
2010-01-05  1136.52  19.35
2010-01-06  1137.14  19.16
2010-01-07  1141.69  19.06
2010-01-08  1144.98  18.13
2010-01-11  1146.98  17.55
2010-01-12  1136.22  18.25
2010-01-13  1145.68  17.85
2010-01-14  1148.46  17.63
2010-01-15  1136.03  17.91
2010-01-18      NaN    NaN
2010-01-19  1150.23  17.58
2010-01-20  1138.04  18.68
2010-01-21  1116.48  22.27
2010-01-22  1091.76  27.31
2010-01-25  1096.78  25.41
2010-01-26  1092.17  24.55
2010-01-27  1097.50  23.14
2010-01-28  1084.53  23.73
2010-01-29  1073.87  24.62

You can even see in the output above how `Date` is in a different row than `.SPX` and `.VIX`. The date is not a variable anymore. It is an index.

This is a much simpler example than the Zillow data. 